[TypeAlias] Add LowerToHW support for lowering TypeAlias

prithayan opened the pull request:
Lower `firrtl.alias` to `hw.typealias`. 
Create a global `hw.type_scope` and add the `hw.typedecl`s corresponding to all
the `firrtl.alias`s across  the `firrtl.module`s. 
This PR adds a pre-scan to the `LowerToHW` pass to lower all the `firrtl.alias`
types for all the modules, sequentially. This ensures that name conflict
resolution and declaration order of `typedecl` are deterministic. The
sequential pass records all the alias types and their corresponding lowering,
which is accessed during the parallel phase.

uenoku commented on lib/Dialect/FIRRTL/FIRRTLUtils.cpp:
```diff
@@ -822,26 +822,37 @@ circt::firrtl::maybeStringToLocation(StringRef spelling, bool skipParsing,
     return type;
 
+  if (BaseTypeAliasType aliasType =
+          firrtl::type_dyn_cast<BaseTypeAliasType>(firType)) {
```
type_dyn_cast strips outermost type aliases. To access type alias, it's necessary to use normal dyn_cast.

prithayan commented:
@uenoku , I made the type alias lowering deterministic.
I updated the pass to only create `TypeScopeOp` at the `mlir::ModuleOp` level for port types. This ensures that any global scope Type aliases are created sequentially!!
Any other type alias that is created for `HW::HWMOduleOp` body, is always local and hence only executed sequentially.

uenoku commented on test/firtool/type-alias.fir:
Nice!

prithayan commented:
@uenoku, I updated this PR and added a lit test. Will add a few more tests.
I tried to implement a deterministic lowering of type names under parallel execution.

prithayan commented on lib/Conversion/FIRRTLToHW/LowerToHW.cpp:
```diff
@@ -2518,7 +2614,8 @@ FailureOr<Value> FIRRTLLowering::lowerSubaccess(SubaccessOp op, Value input) {
 FailureOr<Value> FIRRTLLowering::lowerSubfield(SubfieldOp op, Value input) {
-  auto resultType = lowerType(op->getResult(0).getType());
+  auto resultType = circuitState.lowerType(
+      op->getResult(0).getType(), builder.getLoc(), theModule.getName());
```
Thanks Hideto, good idea to make it simpler, added the wrapper.

prithayan commented on lib/Dialect/FIRRTL/FIRRTLUtils.cpp:
```diff
@@ -886,26 +886,36 @@ circt::firrtl::maybeStringToLocation(StringRef spelling, bool skipParsing,
+    type = lowerType(aliasType.getInnerType(), loc, getTypeDeclFn);
+    if (!loc)
+      loc = UnknownLoc::get(type.getContext());
+    if (getTypeDeclFn)
```
Yes, hoisted it and switched to your `type_dyn_cast`, to make this work. Tested with the functionref to be null.

prithayan commented on lib/Conversion/FIRRTLToHW/LowerToHW.cpp:
```diff
@@ -265,6 +268,52 @@ struct CircuitLoweringState {
+      return iterT1->second;
+    // Mutex, to update the top level Module, and add the TypeDeclOp inside the
+    // TypeDeclOp.
+    std::lock_guard<std::mutex> lock(typeScopeMutex);
+    // The typescope is created only at the top level module.
+    Block &topLevelBlock = circuitOp->getParentRegion()->getBlocks().back();
```
So, the ports are not lowered in parallel, so name conflict with ports across modules should be fine. Types within a module are lowered sequentially, that should also be fine. I was thinking type name conflict across modules can be resolved by appending the module name to the type. But, you are right, the current implementation only appends the module name on conflict, which can result in non-deterministic result. Did we specify this type name conflict behavior in the spec ?

uenoku replied:
I agree that type alias which appears in ports are fine but I think type aliases in modules are lowered parallely so it looks unsafe to mutate single type decl. 
I guess currently type name conflicts would not happen because type aliases are unique at parser level, and type aliases will be dropped when inner types are mutated.

uenoku replied:
> So, here I used the lock on line 284, that should be make mutating the type decl safe, right ?

Yes, I think this is thread safe but operations order would be non-deterministic.

uenoku commented on lib/Conversion/FIRRTLToHW/LowerToHW.cpp:
```diff
@@ -329,6 +354,79 @@ struct CircuitLoweringState {
+  /// deteministic TypeDecls inside the global TypeScopeOp.
+  struct RecordTypeAlias {
+
+    RecordTypeAlias(CircuitOp &c) : circuitOp(c) {}
```
```suggestion
    RecordTypeAlias(CircuitOp c) : circuitOp(c) {}
```

uenoku commented on lib/Conversion/FIRRTLToHW/LowerToHW.cpp:
```diff
@@ -485,6 +583,14 @@ void FIRRTLModuleLowering::runOnOperation() {
           state.oldToNewModuleMap[&op] = loweredMod;
           modulesToProcess.push_back(module);
+          // Lower all the alias types.
+          module.walk([&](Operation *op) {
+            for (auto res : op->getResults()) {
+              if (auto aliasType =
+                      type_dyn_cast<BaseTypeAliasType>(res.getType()))
+                state.lowerType(aliasType, op->getLoc());
+            }
+          });
```
Walking entire module could be heavy so it would be nice to paralelly accumulate type aliases (without constructing type decl) in the future.

prithayan replied:
That is a good idea, but as accumulation in parallel will result in non-deterministic order, we need to order them before lowering. Will try it in a separate PR.

uenoku commented on lib/Conversion/FIRRTLToHW/LowerToHW.cpp:
```diff
@@ -265,6 +266,19 @@ struct CircuitLoweringState {
 
+  /// For a given Type Alias, return the corresponding AliasType. Create and
+  /// record the AliasType, if it doesn't exist.
+  hw::TypeAliasType getTypeAlias(Type rawType, BaseTypeAliasType firAliasType,
```
nit: Can we make argument order consistent with addTypedecl?

prithayan replied:
Thanks for spotting this, fixed.

uenoku commented on include/circt/Dialect/FIRRTL/FIRRTLUtils.h:
```diff
@@ -232,8 +232,12 @@ mapBaseTypeNullable(FIRRTLType type,
+/// cannot be lowered. The optional function is required to specify how to lower
+/// AliasTypes.
+Type lowerType(
+    Type type, std::optional<Location> loc = {},
```
It is necessary to use `std::optional`? Can we just use nullptr?

prithayan replied:
I think Location can't be null, we need to use `optional`

uenoku reviewed (approved):
Awesome, thank you for working on this. I left several nit comments but otherwise LGTM if there is no significant compile time regression.

prithayan commented:
Thanks Hideto, addressed your feedback. 
I ran on a few internal designs, but Didn't notice any increase in `LowerToHW` time. Will try to parallelize the walk in a followup PR.
